### DISCUSSION TITLE
Correct way to center the signal by channel and by trial/event

### DISCUSSION THREAD

**Discussion by lnalborczyk** [bnicenboim/eeguana#219] [Q&A] [answered]
Hi,

let me first thank you (again) for developing and maintaining the package, it is super helpful!

I have an `eeg_lst` object with 200 events (as displayed below).

```
# EEG data:

# Signal table:
        .id .sample    OO   APB
     1:   1    -249 -12.7 -78.4
     2:   1    -248 -16.4 -76.9
     3:   1    -247 -18.1 -77.0
     4:   1    -246 -18.8 -78.8
     5:   1    -245 -20.2 -80.0
    ---                        
100196: 200     247  57.8 -78.9
100197: 200     248  57.4 -78.3
100198: 200     249  55.8 -77.2
100199: 200     250  54.2 -76.5
100200: 200     251  52.9 -75.7

# Events table:
     .id    .type .description .initial .final .channel
  1:   1 Response         R128        1      1     <NA>
  2:   2 Response         R128        1      1     <NA>
  3:   3 Response         R128        1      1     <NA>
  4:   4 Response         R128        1      1     <NA>
  5:   5 Response         R128        1      1     <NA>
 ---                                                   
196: 196 Response         R128        1      1     <NA>
197: 197 Response         R128        1      1     <NA>
198: 198 Response         R128        1      1     <NA>
199: 199 Response         R128        1      1     <NA>
200: 200 Response         R128        1      1     <NA>

# Segments table:
     .id                 .recording segment description
  1:   1 data/S01/S01_lad_0014.vhdr       1        R128
  2:   2 data/S01/S01_lad_0014.vhdr       2        R128
  3:   3 data/S01/S01_lad_0014.vhdr       3        R128
  4:   4 data/S01/S01_lad_0014.vhdr       4        R128
  5:   5 data/S01/S01_lad_0014.vhdr       5        R128
 ---                                                   
196: 196 data/S01/S01_lad_0014.vhdr     196        R128
197: 197 data/S01/S01_lad_0014.vhdr     197        R128
198: 198 data/S01/S01_lad_0014.vhdr     198        R128
199: 199 data/S01/S01_lad_0014.vhdr     199        R128
200: 200 data/S01/S01_lad_0014.vhdr     200        R128
```

I would like to center the signal **by trial** and by channel. I have tried using the `eeg_baseline()` function (which [seems to center the signal](https://github.com/bnicenboim/eeguana/blob/1a2a0381d5e7707e2a49895acd6b44163870d8f8/R/channel_functions.R) on each trial/event), but here I guess it does not do what I want because it considers all previous timesteps when centering. When computing the mean of the signal by trial and by channed with the following code, it indeed appears not to be 0.

```
data_segments %>%
    as_tibble() %>%
    group_by(.id, .key) %>%
    summarise(m = mean(.value) ) %>%
    ungroup()
```

What would be the correct way to center the signal by channel and by trial? By correct I mean that the mean of the signal in each trial and each channel is effectively 0.

Thank you for your help!

Ladislas

**Comment by bnicenboim** [comment, marked as answer]
Hi! Thanks for the kind words!

Yeah, good point. `eeg_baseline` only baselines using the signal before the start of the trial (the time 0). That's how I've always done it and seen it. But I could make `eeg_baseline` more flexible. Could you open an issue and provide a reference on a paper where it's done the way you want it?

In any case, what you want to do (if I understood it right), it's relatively straight forward to do:

```
data_segments <- data_faces_ERPs %>% eeg_group_by(.id) %>%
  eeg_mutate(across_ch(~.x - mean(.x)))
```

`across_ch` is like dplyr's across but by default across all the channels of the object, so for every channel in every segment, you take it and subtract the mean. That's what you wanted to do, right?

Best regards,
Bruno

**Reply by lnalborczyk** [reply]
Hi,

great, thank you for your prompt reply and for the code, it works like a charm!

Yeah, I do not think it is necessary to modify the function. I have used `eeg_baseline()` not to "baseline" the signal,  but rather to center it to avoid a potential offset, and I wanted to do that per trial (rather than on the whole channel), to take into account a potential drift throughout the experiment (I am working with EMG signals (MEPs)).

Anyway, the code you have shared does exactly what I wanted to do :)

Thank you again for your help!

Ladislas
